relations/jf/src/deposit.rs

```diff
@@ -1,85 +1,67 @@
-use ark_ff::Zero;
-use jf_primitives::circuit::rescue::RescueNativeGadget;
-use jf_relation::{Circuit, PlonkCircuit};
+use jf_relation::PlonkCircuit;
 
 use crate::{
-    shielder_types::{convert_hash, Note, Nullifier, TokenAmount, TokenId, Trapdoor},
-    CircuitField, Marshall, PlonkResult, Relation,
+    note::{NoteGadget, NoteType, SourcedNote},
+    shielder_types::{Note, Nullifier, TokenAmount, TokenId, Trapdoor},
+    CircuitField, PlonkResult, PublicInput, Relation,
 };
 
-#[derive(Copy, Clone, Eq, PartialEq, Hash, Default, Debug)]
 pub struct DepositRelation {
-    public: DepositPublicInput,
-    private: DepositPrivateInput,
+    deposit_note: SourcedNote,
 }
 
-#[derive(Copy, Clone, Eq, PartialEq, Hash, Default, Debug)]
+impl Default for DepositRelation {
+    fn default() -> Self {
+        Self::new(Default::default(), Default::default())
+    }
+}
+
+impl DepositRelation {
+    pub fn new(public: DepositPublicInput, private: DepositPrivateInput) -> Self {
+        Self {
+            deposit_note: SourcedNote {
+                note: public.note,
+                token_id: public.token_id,
+                token_amount: public.token_amount,
+                trapdoor: private.trapdoor,
+                nullifier: private.nullifier,
+                note_type: NoteType::Deposit,
+            },
+        }
+    }
+}
+
+#[derive(Default)]
 pub struct DepositPublicInput {
     pub note: Note,
     pub token_id: TokenId,
     pub token_amount: TokenAmount,
 }
 
-impl Marshall for DepositPublicInput {
-    fn marshall(&self) -> Vec<CircuitField> {
-        vec![
-            convert_hash(self.note),
-            self.token_id.into(),
-            self.token_amount.into(),
-        ]
+impl PublicInput for DepositRelation {
+    fn public_input(&self) -> Vec<CircuitField> {
+        self.deposit_note.public_input()
     }
 }
 
-#[derive(Copy, Clone, Eq, PartialEq, Hash, Default, Debug)]
+#[derive(Default)]
 pub struct DepositPrivateInput {
     pub trapdoor: Trapdoor,
     pub nullifier: Nullifier,
 }
 
 impl Relation for DepositRelation {
-    type PublicInput = DepositPublicInput;
-    type PrivateInput = DepositPrivateInput;
-
-    fn new(public_input: Self::PublicInput, private_input: Self::PrivateInput) -> Self {
-        Self {
-            public: public_input,
-            private: private_input,
-        }
-    }
-
     fn generate_subcircuit(&self, circuit: &mut PlonkCircuit<CircuitField>) -> PlonkResult<()> {
```

Comment:
Can we move the ownership of the `circuit` to the method, mutate it internally and return as a result? I think it's cleaner and less bug-prone.

Reply:
`&mut` is more jellyfish-consistent; and also this will be a problem for generating gadgets that return some variable bindings - we would have to return additionally circuit there